### DISCUSSION TITLE
Protected Mode and Error Handling

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#385] [Show and tell]
Here is a very handy link to allow us to make our panels less likely to crash and more importantly less likely to crash the host.

https://www.tutorialspoint.com/lua/lua_error_handling.htm

I think we should use assert and pcall a lot more often in our code.  LUA doesn't seem to do everything in protected mode by default, and VST's don't seem to run in protected mode within VST hosts.  So an error in our code can crash our host.

pcall allows any LUA function to run in protected mode and allows us to throw custom errors, so as a worst case, it should just stop the function from running without crashing your LUA instance and more importantly the host.

**Comment by Ted-juh** [comment]
Very helpful. As far as I could tell, Lua 5.1 has error handling (or pcall) as well. It might differ from version 5.4 which the link indicates at.

**Reply by bijlevel** [reply]
Congrats for solving your issue and indeed useful information!

**Reply by Godlike-Productions** [reply]
I've tested this now, and can confirm that it works at stopping crashes in 5.6.23

**Reply by dobo365** [reply]
> I've tested this now, and can confirm that it works at stopping crashes in 5.6.23

Hi Jason, can you precise how you implemented this? I'm trying my first VST3 with the last 5.6.23 and it crashed Cakewalk directly. Hahaha...
Do you use pcall or xpcall when calling any method from any place of the code?
At panel start I call different Init methods. Shall I call them using pcall?
If not using pcall, do you have some clever way to identify the potential issue?
I will produce standalone and VST2 of the same panel and see if I get the same issues (should have done that first in fact but was eager to try the VST3)
Thx man!

**Reply by Godlike-Productions** [reply]
It depends a little on what I'm doing.  I've used both Assert and xpCall.

All you do is that instead of just calling is like runFunction(a,b), you just go xpCall(runFunction(a,b),"Custom Error Message")

The difference is really in the error message returned.  Assert automatically displays the lua error message, as well as whatever custom error message you put in.  xpCall just returns your custom error message.  If you want to see the internal lua error, you need to build the errorhandler as per the last example at https://www.tutorialspoint.com/lua/lua_error_handling.htm

I'm not sure you can directly call the init method in protected space.  You may need to call an init method that then calls a second function via xpCall

**Reply by dobo365** [reply]
Thx. I'll give it a try!

**Comment by unityconstruct** [comment]
**_Thanks for posting this!!.. _**
_I was parsing the lua ref and found `pcall` when looking for try/catch syntax. Initially couldn't figure it out and bailed on the matter until I saw that you successfully implemented. Worked it out in quick order with this fresh injection of hope...Without something to prevent the Ctrlr host from crashing on every attempt, reverse engineering the API and/or working out exploratory ideas has been a nightmare - one that I'm not inclined to complain about given its open source and delivers incredible power._ 

Confirming **pcall(func())** catches the constant lua-induced Ctrlr pukes.
- About to look at `Assert`
- `xpcall` did not catch exceptions for me

## current design pattern I'm using 
_( not sure if the console output has value in staying )_
```lua
doSomething = function()
 if pcall(doSomethingTry) then
	console("Success")
 else
   console("Fail")
 end
end

doSomethingTry = function()
  -- something smart
end
```

Decided to check here before tabling lua for the time being and just using the poo-brain fields for everything until a brick-wall was found implementing something that needed a more complex handler. Thanks again!

**Reply by Godlike-Productions** [reply]
Thanks for your updates here.  I've had mixed success with xpcall, and haven't managed to find a pattern, but may revert to pcall now if it is more consistent.

It's most necessary in 5.6.23 which just unceremoniously crashes.  The 6.0.xx version, 5.439 and 5.3.201 all have much better error handling during runtime.  But 5.6.23 is the one that we can compile for the VST3 issues.  I'm looking forward to trying the improvements to the community efforts here.  5.6.23 has a number of issues such as no canvas refresh in the VST3 and now not loading into Ableton Live.
